File: TwoPointers/15ThreeSum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums = sorted(nums)
        for i in range(len(nums)):

            if i > 0 and nums[i]==nums[i-1]:
                continue    
            l = i + 1
            r = len(nums)-1
            while l<r: 
                treeSum = nums[i] +nums[l]+nums[r]
                if treeSum > 0 :
                    r -= 1
                    #while nums[r]==nums[r+1] and l < r:
                    #   r -= 1 
                elif treeSum < 0 :
                    l += 1
                    #while nums[l]==nums[l-1] and l < r:
                    #    l += 1
                else :
                    res.append([nums[i],nums[l],nums[r]])
                    l += 1
                    while nums[l]==nums[l-1] and l < r:
                        l += 1

        return res

File: TwoPointers/15ThreeSum.cpp
#include<algorithm>
#include<vector>
class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums) {
        std::vector<std::vector<int>> result;
        std::sort(nums.begin(), nums.end());

        for(int i = 0; i < nums.size(); i++){
            
            if(i > 0 && nums[i] == nums[i-1]) continue;

            int l = i + 1;
            int r = nums.size()-1;
            while(l < r){
                int threeSum = nums[i] + nums[l] + nums[r];

                if(threeSum > 0){
                    r -= 1;
                }
                else if(threeSum < 0){
                    l += 1;
                }
                else {
                    result.push_back({nums[i],nums[l],nums[r]});
                    l += 1;
                    while( nums[l] == nums[l-1] && l < r) l += 1;
                }
            }
        }
        return result;
    }
};
